CRM/Contribute/Form/Contribution/Confirm.php

```diff
@@ -1929,6 +1934,18 @@ protected function processFormSubmission($contactID) {
     // fix currency ID
     $this->_params['currencyID'] = CRM_Core_Config::singleton()->defaultCurrency;
 
+    // CRM-18854
+    if (CRM_Utils_Array::value('adjust_recur_start_date', $this->_values)) {
+      $pledgeBlock = CRM_Pledge_BAO_PledgeBlock::getPledgeBlock($this->_id);
+      if (CRM_Utils_Array::value('start_date', $this->_params) || !CRM_Utils_Array::value('is_pledge_start_date_visible', $pledgeBlock)) {
+        $pledgeStartDate = CRM_Utils_Array::value('start_date', $this->_params, NULL);
+        $this->_params['receive_date'] = CRM_Pledge_BAO_Pledge::getPledgeStartDate($pledgeStartDate, $pledgeBlock);
+        $recurParams = CRM_Pledge_BAO_Pledge::buildRecurParams($this->_params);
+        $this->_params = array_merge($this->_params, $recurParams);
```

Comment:
Right - you obviously added it because something didn't work without it there & so it seemed like the easiest way to get it working - but it's probably not a robust way - so my question is if those 2 lines are not there - where does the problem occur. I presume that 'somewhere later on it wants the value 'contributionRecurID' in order to do 'something'. 

These lines here https://github.com/civicrm/civicrm-core/blob/4.7.9-rc/CRM/Contribute/BAO/Contribution/Utils.php#L123 are also enabling 'something to happen somewhere' - but are not themselves the somewhere. So to fix this without increasing the code fragility we need to figure out what value is actually required where.

This function is considered to be highly toxic but also critical - so any changes to it should attempt to actively reduce it's toxicity & increase code coverage - and you have to assume that before planning to touch it.

If I were going to tackle this I would start by adding a test based on 

api_v3_ContributionPageTest::testSubmitMembershipPriceSetPaymentPaymentProcessorRecurDelayed 

(but taking the mail checking code from another test in the class - e.g testSubmitMembershipBlockIsSeparatePaymentZeroDollarsPayLaterWithEmail()  )

And make sure that when the form is submitted all the correct entities are created & the mail has the correct contents. Then I would remove the 2 lines & see if the test still passes & if not focus on a better way to get the test passing again